### ISSUE TITLE
[mlir][emitc] Parenthesization issue in TranslateToCpp

### ISSUE THREAD

**Issue by cferry-AMD** [llvm/llvm-project#93470]
This MLIR: 

```mlir
module {
  func.func @shift() {
    %0 = "emitc.constant"() <{value = 160 : ui32}> : () -> ui32
    %1 = "emitc.constant"() <{value = 7 : ui32}> : () -> ui32
    %2 = emitc.cmp lt, %0, %1 : (ui32, ui32) -> i1
    %3 = emitc.expression : ui32 {
      %4 = emitc.bitwise_right_shift %0, %1 : (ui32, ui32) -> ui32
      %5 = emitc.conditional %2, %4, %0 : ui32
      emitc.yield %5 : ui32
    }
    %6 = emitc.cast %3 : ui32 to i32
    return
  }
}
```

gets translated into this C++ code using `mlir-translate --mlir-to-cpp`:
```c++
void shift() {
  uint32_t v1 = 160;
  uint32_t v2 = 7;
  bool v3 = v1 < v2;
  int32_t v4 = (int32_t) v3 ? v1 >> v2 : v1;
  return;
}
```

where we would expect:
```c++
  int32_t v4 = (int32_t)(v3 ? v1 >> v2 : v1);
```

or
```c++
  uint32_t v4 = v3 ? v1 >> v2 : v1;
  int32_t v5 = (int32_t) v4;
```

The `CppEmitter` needs to emit parentheses around casted values. Here, it doesn't and the cast happens on the predicate, which is incorrect. Fortunately, this looks pretty easy to fix :)

**Comment by mgehre-amd** [comment]
fyi @marbre, @simon-camp, @christopherbate

**Comment by simon-camp** [comment]
This affects all users of expression op results.

I found another issue while playing around. Chained operators with the same precedence are not parenthesized and therefore potentially evaluated in the wrong order.
```mlir
func.func @z(%v1 : ui32, %v2 : ui32, %v3 : ui32) -> ui32 {
    %1 = emitc.expression : ui32 {
      %4 = emitc.mul %v1, %v2 : (ui32, ui32) -> ui32
      %5 = emitc.div %v3, %4 : (ui32, ui32) -> ui32
      emitc.yield %5 : ui32
    }

    return %1 : ui32
  }
```

is incorrectly emitted as
```C
uint32_t z(uint32_t v1, uint32_t v2, uint32_t v3) {
  return v3 / v1 * v2;  // should be v3 / (v1 * v2)
}
```
I will send a patch for this error tomorrow.

**Comment by simon-camp** [comment]
First try of a fix is ready for review here #93691.
